browser/net/brave_ad_block_tp_network_delegate_helper.cc

```diff
@@ -115,15 +108,28 @@ void OnShouldBlockAdResult(const ResponseCallback& next_callback,
   next_callback.Run();
 }
 
-void ShouldBlockAdWithOptionalCname(
-    scoped_refptr<base::SequencedTaskRunner> task_runner,
-    const ResponseCallback& next_callback,
-    std::shared_ptr<BraveRequestInfo> ctx,
-    const base::Optional<std::string> cname) {
+void UseCnameResult(scoped_refptr<base::SequencedTaskRunner> task_runner,
+                    const ResponseCallback& next_callback,
+                    std::shared_ptr<BraveRequestInfo> ctx,
+                    const base::Optional<std::string> cname) {
   DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
-  task_runner->PostTaskAndReply(
-      FROM_HERE, base::BindOnce(&ShouldBlockAdOnTaskRunner, ctx, cname),
-      base::BindOnce(&OnShouldBlockAdResult, next_callback, ctx));
+
+  if (cname.has_value() && ctx->request_url.host() != *cname &&
+      !cname->empty()) {
+    GURL::Replacements replacements;
+    replacements.SetHost(cname->c_str(),
+                         url::Component(0, static_cast<int>(cname->length())));
+    const GURL canonical_url = ctx->request_url.ReplaceComponents(replacements);
+
+    task_runner->PostTaskAndReply(
+        FROM_HERE,
+        base::BindOnce(&ShouldBlockRequestOnTaskRunner, ctx,
+                       base::make_optional<GURL>(canonical_url)),
+        base::BindOnce(&OnShouldBlockUncloakedRequestResult, next_callback,
+                       ctx));
+  } else {
+    next_callback.Run();
+  }
 }
 
 class AdblockCnameResolveHostClient : public network::mojom::ResolveHostClient {
```

Comment:
btw this can be inherited from `network::ResolveHostClientBase` to save few boilerplate lines

Reply:
Good to know, but I'll leave it for a followup since I'm not really touching it in this PR